### PR TITLE
 【BIT】转换能力补齐：paddle.nn.functional.[batch_norm bilinear channel_shuffle flashmask_attention fractional_max_pool[2d 3d]]

### PR DESCRIPTION
**Pull request by cangtianhuang**
### 1. [paddle.nn.functional.batch_norm](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/nn/functional/batch_norm_cn.html)

查询 [torch.nn.functional.batch_norm 与 paddle.nn.functional.batch_norm 对照表](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/guides/model_convert/convert_from_pytorch/api_difference/functional/torch.nn.functional.batch_norm.html) ，已有详细对比，发现是 `paddle 参数更多` ，因此为其注册 BatchNormRule 类。

paddle 相较于 torch 多了 `data_format` 参数，表示输入和输出的数据格式，可以是 "NCHW" 和 "NHWC"，而 torch 默认是 "NCHW"，因此需首先使用 permute 方法调整维度顺序，计算完毕后再使用 permute 调整回来。

此外，momentum 含义不同，值的大小 Paddle = 1 - PyTorch，需要修改；running_mean 与 running_var 的 Tensor 需要主动关闭梯度传播。具体操作见 Rule 代码。

### 2. [paddle.nn.functional.bilinear](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/nn/functional/bilinear_cn.html)

查询 [torch.nn.functional.bilinear 与 paddle.nn.functional.bilinear 对照表](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/guides/model_convert/convert_from_pytorch/api_difference/functional/torch.nn.functional.bilinear.html) ，已有详细对比，发现是 `参数用法不一致` 。paddle 的 bias 参数多了一个大小为 1 的维度，可以通过仅撰写配置解决。

### 3. [paddle.nn.functional.channel_shuffle](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/nn/functional/channel_shuffle_cn.html)

查询 [torch.nn.functional.channel_shuffle 与 paddle.nn.functional.channel_shuffle 对照表](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/guides/model_convert/convert_from_pytorch/api_difference/functional/torch.nn.functional.channel_shuffle.html) ，已有详细对比，发现是 `paddle 参数更多` ，同样是 paddle 相较于 torch 多了 `data_format` 参数，因此为其注册了更为通用的 DataFormatRule 类，统一解决 permute 问题。

### 4. paddle.nn.functional.flashmask_attention

此 API 在官方的中英文文档中均未查询到。查看 paddle 的 [flashmask_attention 源码](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/nn/functional/flash_attention.py#L405) 后，发现是新 API （6-8月前被开发）。查阅资料发现 torch 中没有对应 API，难以做精度对比，因此不做处理。

### 5. [paddle.nn.functional.fractional_max_pool2d](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/nn/functional/fractional_max_pool2d_cn.html)

查询 [torch.nn.functional.fractional_max_pool2d 与 paddle.nn.functional.fractional_max_pool2d 对照表](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/guides/model_convert/convert_from_pytorch/api_difference/functional/torch.nn.functional.fractional_max_pool2d.html) ，已有详细对比，发现是 `torch 参数更多` ，但是大部分参数用法不一致，因此为其注册 FractionalMaxPoolRule 类。

总体来看，paddle 的参数允许更丰富的表现形式，而 torch 的参数类型更保守。`random_u` 参数 paddle 中为 float 或 None， torch 对应 `_random_samples` 参数，为列表式的 Tensor，需要转写；`kernel_size` 参数 paddle 支持为 None，需要手动通过 x 与 output_size 推导形状并转换为 tuple；`output_size` 参数 paddle 支持其中元素为 None，也需要推导。具体操作见 Rule 代码。

在实际测试中发现，因为有 `random_u` 的存在，api 的输出具有随机性，因此判断其无法进行精度对比（但支持配置转换）。

### 6. [paddle.nn.functional.fractional_max_pool3d](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/nn/functional/fractional_max_pool3d_cn.html)

查询 [torch.nn.functional.fractional_max_pool3d 与 paddle.nn.functional.fractional_max_pool3d 对照表](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/guides/model_convert/convert_from_pytorch/api_difference/functional/torch.nn.functional.fractional_max_pool3d.html) ，已有详细对比，发现是 `torch 参数更多` ，但是大部分参数用法不一致，与 ractional_max_pool2d 共用 FractionalMaxPoolRule 类。后续处理同 fractional_max_pool2d 。

## 其他修改

1. 为 accuracy.py 添加 traceback.print_exc() ，帮助同学们定位 Rule 中的报错。
2. 修复了 need_check_grad() 方法，可读性更高。（已回退至之前版本，避免引入新问题，保留了注释）

## 测试结果

paddle.nn.functional.[flashmask_attention fractional_max_pool[2d 3d]] 的配置由于随机性全部移入 api_config_paddleonly_1.txt； paddle.nn.functional.batch_norm 部分配置未通过精度对比，不能判断是什么原因造成，移入 api_config_accuracy_error.txt，并在 fresh_report 中报告；其余配置均移入 api_config_support2torch_1.txt 。

image

### CODE REVIEW

**Review by yuwu46 — APPROVED** [review]
LGTM

**Review by wanghuancoder — APPROVED** [review]
LGTM
